Services/Devices.cs

```diff
@@ -134,78 +140,82 @@ public IDeviceClient GetClient(Device device, IoTHubProtocol protocol, IScriptIn
+
+            return new Device(
+                this.PrepareDeviceObject(deviceId, this.fixedDeviceKey),
+                this.ioTHubHostName);
+        }
+
+        // Get the device from the registry
         public async Task<Device> GetAsync(string deviceId)
         {
-            this.SetupHub();
+            this.CheckSetup();
 
             this.log.Debug("Fetching device from registry", () => new { deviceId });
 
             Device result = null;
-            var now = DateTimeOffset.UtcNow;
+            var start = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
 
             try
             {
-                var device = await this.GetRegistry().GetDeviceAsync(deviceId);
+                Azure.Devices.Device device = await this.registry.GetDeviceAsync(deviceId);
                 if (device != null)
                 {
                     result = new Device(device, this.ioTHubHostName);
                 }
                 else
                 {
-                    this.log.Debug("Device not found", () => new { deviceId });
+                    var timeSpentMsecs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - start;
+                    this.log.Debug("Device not found", () => new { timeSpentMsecs, deviceId });
                 }
             }
+            catch (Exception e) when (e is TaskCanceledException || e.InnerException is TaskCanceledException)
+            {
+                var timeSpentMsecs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - start;
+                this.log.Error("Get device task timed out", () => new { timeSpentMsecs, deviceId, e.Message });
+                throw new ExternalDependencyException("Get device task timed out", e);
+            }
             catch (Exception e)
             {
-                if (e.InnerException != null && e.InnerException.GetType() == typeof(TaskCanceledException))
-                {
-                    var timeSpent = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - now.ToUnixTimeMilliseconds();
-                    this.log.Error("Get device task timed out", () => new { timeSpent, deviceId, e.Message });
-                    throw;
-                }
-
-                this.log.Error("Unable to fetch the IoT device", () => new { deviceId, e });
+                var timeSpentMsecs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - start;
+                this.log.Error("Unable to fetch the IoT device", () => new { timeSpentMsecs, deviceId, e });
                 throw new ExternalDependencyException("Unable to fetch the IoT device");
             }
 
             return result;
         }
 
-        /// <summary>
-        /// Register the new device
-        /// </summary>
+        // Register a new device
         public async Task<Device> CreateAsync(string deviceId)
         {
-            this.SetupHub();
-            var now = DateTimeOffset.UtcNow;
+            this.CheckSetup();
+            var start = DateTimeOffset.UtcNow;
```

Comment:
for consistency sake, can we change this to start = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds here and then just use start var below in exception block